### PR TITLE
Seed repo with pyro.generic

### PR DESCRIPTION
**Pull request by neerajprad**
This seeds the module with existing code from `pyro.generic`. Depending on the future plans for `pyro.generic` module, we can either delete that and maintain this repo instead, or have some subset of that reside here for generic interface testing. I would prefer the former so that we don't need to duplicate code across both the repositories, but if we have a future requirement that needs generic to depend on pytorch or pyro, then it would make sense to only keep the backend agnostic subset here. 

Note the following:
 - There are no dependencies for the pyro-api package, and this can be easily imported by other libraries.
 - I have added `funsor`, `pyro-ppl`, and `numpyro` as test dependencies. I am currently pinning them to some latest dev commits to run some interface tests in `test`. These dependencies can be later changed to the pypi versions after the next release for funsor, pyro and numpyro, if we like.

### CODE REVIEW

File: pyro_api/__init__.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+from pyro_api.dispatch import distributions, handlers, infer, ops, optim, pyro, pyro_backend
```
Comment (fritzo) [inline pyro_api/__init__.py]: pep8 [discourages](https://www.python.org/dev/peps/pep-0008/#package-and-module-names) use of underscores in package names. Do you think it's worth renaming this to something like `pyroapi`, or should we disregard pep8?

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+setup(
+    name='pyro-api',
+    version=version,
+    description='Pyro API for generic model dispatch.',
+    packages=find_packages(include=['pyro_api', 'pyro_api.*']),
+    url='https://github.com/pyro-ppl/pyro-api',
+    author='Uber AI Labs',
+    author_email='npradhan@uber.com',
+    install_requires=[],
```
Comment (fritzo) [inline setup.py]: :100:

File: pyro_api/__init__.py (hunk 1 above)
Comment (neerajprad) [reply pyro_api/__init__.py]: It was looking a bit odd for sure. `pyroapi` seems good to me, happy to change to that instead.

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+except Exception as e:
+    sys.stderr.write('Failed to convert README.md to rst:\n  {}\n'.format(e))
+    sys.stderr.flush()
+    long_description = ''
+
+
+setup(
+    name='pyro-api',
+    version=version,
+    description='Pyro API for generic model dispatch.',
```
Comment (fritzo) [inline setup.py]: nit: maybe mention 'backend', say as in
```
Generic API for dispatch to Pyro backends.
```

File: test/test_dispatch.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+from pyro_api import handlers, infer, pyro, pyro_backend
+from pyro_api.testing import MODELS
+
+pytestmark = pytest.mark.stage('unit')
+
+
+@pytest.mark.filterwarnings("ignore", category=UserWarning)
+@pytest.mark.parametrize('model', MODELS)
+@pytest.mark.parametrize('backend', ['pyro', 'numpy'])
+def test_mcmc_interface(model, backend):
```
Comment (fritzo) [inline test/test_dispatch.py]: Just a reminder that any code in this file is not distributed in the package, and therefore is not usable by other libraries. Issue number 1 should be "move all testing code into the main library".

**Comment by fritzo** [comment]
I agree we should delete pyro/generic/ and instead make Pyro depend on this library.

File: pyro_api/__init__.py (hunk 1 above)
Comment (fritzo) [reply pyro_api/__init__.py]: Yeah, let's make the package named `pyroapi`.

File: setup.py (hunk 3 above)
Comment (neerajprad) [reply setup.py]: Thanks, that's much better.

File: test/test_dispatch.py (hunk 4 above)
Comment (neerajprad) [reply test/test_dispatch.py]: Done.

**Comment by fritzo** [comment]
Feel free to merge. Whereas our pyro-ppl policy is "author may not merge", in funsor we find it faster to allow author to merge. What is your policy in Numpyro?

**Comment by neerajprad** [comment]
> Feel free to merge. Whereas our pyro-ppl policy is "author may not merge", in funsor we find it faster to allow author to merge (after approval). What is your policy in Numpyro?

We have the same policy in numpyro as funsor. :)

**Pull request merged**